### ISSUE TITLE
fmt 10

### ISSUE THREAD

**Issue by dschwoerer** [boutproject/BOUT-dev#2729]
BOUT++ FTBFS with fmt 10:
```
In file included from /builddir/build/BUILD/BOUT++-v5.0.0.dev10626+gd935fa1ea/include/bout/boutexception.hxx:15,
                 from /builddir/build/BUILD/BOUT++-v5.0.0.dev10626+gd935fa1ea/include/bout/solver.hxx:42,
                 from /builddir/build/BUILD/BOUT++-v5.0.0.dev10626+gd935fa1ea/include/bout/invert/laplacexy.hxx:72,
                 from /builddir/build/BUILD/BOUT++-v5.0.0.dev10626+gd935fa1ea/src/invert/laplacexy/laplacexy.cxx:5:
/usr/include/fmt/core.h: In instantiation of 'constexpr fmt::v10::detail::value<Context> fmt::v10::detail::make_value(T&&) [with Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; T = const KSPConvergedReason&]':
/usr/include/fmt/core.h:1711:29:   required from 'constexpr fmt::v10::detail::value<Context> fmt::v10::detail::make_arg(T&&) [with bool IS_PACKED = true; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; type <anonymous> = fmt::v10::detail::type::custom_type; T = const KSPConvergedReason&; typename std::enable_if<IS_PACKED, int>::type <anonymous> = 0]'
/usr/include/fmt/core.h:1827:77:   required from 'constexpr fmt::v10::format_arg_store<Context, Args>::format_arg_store(T&& ...) [with T = {const KSPConvergedReason&}; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; Args = {KSPConvergedReason}]'
/usr/include/fmt/core.h:1844:31:   required from 'constexpr fmt::v10::format_arg_store<Context, typename std::remove_cv<typename std::remove_reference<T>::type>::type ...> fmt::v10::make_format_args(T&& ...) [with Context = basic_format_context<appender, char>; T = {const KSPConvergedReason&}]'
/usr/include/fmt/core.h:2817:44:   required from 'std::string fmt::v10::format(format_string<T ...>, T&& ...) [with T = {const KSPConvergedReason&}; std::string = std::__cxx11::basic_string<char>; format_string<T ...> = basic_format_string<char, const KSPConvergedReason&>]'
/builddir/build/BUILD/BOUT++-v5.0.0.dev10626+gd935fa1ea/include/bout/boutexception.hxx:27:34:   required from 'BoutException::BoutException(const S&, const Args& ...) [with S = char [42]; Args = {KSPConvergedReason}]'
/builddir/build/BUILD/BOUT++-v5.0.0.dev10626+gd935fa1ea/src/invert/laplacexy/laplacexy.cxx:1202:76:   required from here
/usr/include/fmt/core.h:1691:7: error: static assertion failed: Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt
 1691 |       formattable,
      |       ^~~~~~~~~~~
/usr/include/fmt/core.h:1691:7: note: 'formattable' evaluates to false
gmake[2]: *** [CMakeFiles/bout++.dir/build.make:803: CMakeFiles/bout++.dir/src/invert/laplacexy/laplacexy.cxx.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
In file included from /builddir/build/BUILD/BOUT++-v5.0.0.dev10626+gd935fa1ea/include/bout/boutexception.hxx:15,
                 from /builddir/build/BUILD/BOUT++-v5.0.0.dev10626+gd935fa1ea/include/bout/assert.hxx:20,
                 from /builddir/build/BUILD/BOUT++-v5.0.0.dev10626+gd935fa1ea/src/invert/laplacexy2/laplacexy2.cxx:5:
/usr/include/fmt/core.h: In instantiation of 'constexpr fmt::v10::detail::value<Context> fmt::v10::detail::make_value(T&&) [with Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; T = const KSPConvergedReason&]':
/usr/include/fmt/core.h:1711:29:   required from 'constexpr fmt::v10::detail::value<Context> fmt::v10::detail::make_arg(T&&) [with bool IS_PACKED = true; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; type <anonymous> = fmt::v10::detail::type::custom_type; T = const KSPConvergedReason&; typename std::enable_if<IS_PACKED, int>::type <anonymous> = 0]'
/usr/include/fmt/core.h:1827:77:   required from 'constexpr fmt::v10::format_arg_store<Context, Args>::format_arg_store(T&& ...) [with T = {const KSPConvergedReason&}; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; Args = {KSPConvergedReason}]'
/usr/include/fmt/core.h:1844:31:   required from 'constexpr fmt::v10::format_arg_store<Context, typename std::remove_cv<typename std::remove_reference<T>::type>::type ...> fmt::v10::make_format_args(T&& ...) [with Context = basic_format_context<appender, char>; T = {const KSPConvergedReason&}]'
/usr/include/fmt/core.h:2817:44:   required from 'std::string fmt::v10::format(format_string<T ...>, T&& ...) [with T = {const KSPConvergedReason&}; std::string = std::__cxx11::basic_string<char>; format_string<T ...> = basic_format_string<char, const KSPConvergedReason&>]'
/builddir/build/BUILD/BOUT++-v5.0.0.dev10626+gd935fa1ea/include/bout/boutexception.hxx:27:34:   required from 'BoutException::BoutException(const S&, const Args& ...) [with S = char [43]; Args = {KSPConvergedReason}]'
/builddir/build/BUILD/BOUT++-v5.0.0.dev10626+gd935fa1ea/src/invert/laplacexy2/laplacexy2.cxx:386:77:   required from here
/usr/include/fmt/core.h:1691:7: error: static assertion failed: Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt
 1691 |       formattable,
      |       ^~~~~~~~~~~
/usr/include/fmt/core.h:1691:7: note: 'formattable' evaluates to false
```

I guess we need a formatter for `KSPConvergedReason`?

**Comment by ZedThree** [comment]
Fix is probably this:

```cpp
    throw BoutException("LaplaceXY2 failed to converge. Reason {} ({:d})",
                        KSPConvergedReasons[reason], static_cast<int>(reason));
```

**Comment by dschwoerer** [comment]
Fixed by: #2742

**Comment by dschwoerer** [comment]
Fixed by: #2764
